c7n/resources/iam.py

```diff
@@ -774,6 +774,120 @@ def _user_mfa_devices(resource):
+    (http://docs.aws.amazon.com/IAM/latest/UserGuide/id_credentials_getting-report.html), but the
+    added benefit of performing less API calls, whereas 'username' will make more API calls, but
+    have a SLA of your cache.
+
+    ... code-block: yaml
+
+     # using a 'credential' filter':
+     - name: iam-only-whitelisted-users
+       resource: iam-user
+       filters:
+         - type: credential
+           key: user
+           op: not-in
+           value:
+             - valid-user-1
+             - valid-user-2
+       actions:
+         - delete
+
+     # using a 'username' filter with 'UserName':
+     - name: iam-only-whitelisted-users
+       resource: iam-user
+       filters:
+         - type: username
+           key: UserName
+           op: not-in
+           value:
+             - valid-user-1
+             - valid-user-2
+       actions:
+         - delete
+
+     # using a 'username' filter with 'Arn':
+     - name: iam-only-whitelisted-users
+       resource: iam-user
+       filters:
+         - type: username
+           key: Arn
+           op: not-in
+           value:
+             - arn:aws:iam:123456789012:user/valid-user-1
+             - arn:aws:iam:123456789012:user/valid-user-2
+       actions:
+         - delete
+
+    """
+
+    schema = type_schema('delete')
+    permissions = ('iam:DeleteLoginProfile',
+                   'iam:ListAccessKeys', 'iam:DeleteAccessKey',
+                   'iam:ListAttachedUserPolicies', 'iam:DetachUserPolicy',
+                   'iam:DeleteUser')
+
+    def process(self, resources):
+        client = local_session(self.manager.session_factory).client('iam')
+
+        for r in resources:
+            self.log.debug('Deleting LoginProfile for user "%s"' % (r['UserName']))
+            try:
+                client.delete_login_profile(UserName=r['UserName'])
+            except ClientError as e:
+                if e.response['Error']['Code'] == 'NoSuchEntity':
+                    self.log.debug('No LoginProfile found to delete for user "%s"' %
+                            r['UserName'])
+                else:
+                    raise
+
+            self.log.debug('Deleting AccessKeys for user "%s"' % r['UserName'])
+            response = client.list_access_keys(UserName=r['UserName'])
+            for access_key in response['AccessKeyMetadata']:
+                client.delete_access_key(UserName=r['UserName'],
+                        AccessKeyId=access_key['AccessKeyId'])
+
+            self.log.debug('Detaching UserPolicies from user "%s"' % (r['UserName']))
+            response = client.list_attached_user_policies(UserName=r['UserName'])
+            for user_policy in response['AttachedPolicies']:
+                client.detach_user_policy(UserName=r['UserName'],
+                        PolicyArn=user_policy['PolicyArn'])
+
+            self.log.debug('Deleting user "%s"' % (r['UserName']))
+            client.delete_user(UserName=r['UserName'])
```

Comment:
if you just delete the user, do you really need to do any of the other activities here, ie a deleted user doesn't have attached user policies or valid access keys i would assume.